### ISSUE TITLE
JavaScript Backend: Double's type is inferred incorrectly

### ISSUE THREAD

**Issue by DoctorRyner** [idris-lang/Idris2#802]
# Steps to Reproduce
1. Download https://github.com/DoctorRyner/idris-js/tree/52a12daa543264a38098e1f0df3e105f5989dbed
2. Run yarn test
3. See the next error:

```
/Users/rustamismailov/tmp/idris/js/build/exec/index.js:22
 return __esPrim_crashExp(invalid cast: + Integer + ' -> ' + Double);
                          ^^^^^^^

SyntaxError: missing ) after argument list
```

# Expected Behavior

There shouldn't be any issues with it

# Observed Behavior

The problem occurs here:

https://github.com/DoctorRyner/idris-js/blob/52a12daa543264a38098e1f0df3e105f5989dbed/src/Js/Array.idr
```idris
xs : Array Double
xs = cons 1 $ singleton 2
```

If you change it to:

```idris
xs : Array Double
xs = cons 1 $ singleton $ the Double 2
```

Then it'll work fine

**Comment by andorp** [comment]
I also found this one, wanted to report it too after finishing what I worked on.

**Comment by andorp** [comment]
There is a good resource on FFI:
https://github.com/rbarreiro/ifui/blob/663e9bc6ea59338cfdea56d3d5b2740a7bf73d84/Ifui/Dom.idr

**Comment by DoctorRyner** [comment]
Oh, I also see that you use

```idris
data DomNode = MkNode AnyPtr
```

To create foreign data, there are at least 2 more ways to do it

```idris
DomNode : Type
```

```idris
DomNode : Type where [external]
```

Can someone explain the difference and the right way to do so?

**Comment by DoctorRyner** [comment]
And btw, this bug doesn't occur when we add .0 to numbers

```Idris
xs : Array Double
xs = cons 1.0 $ singleton 2.0
```

**Comment by DoctorRyner** [comment]
This is a much more critical bug, is there any guesses why it happens?

**Comment by andorp** [comment]
I can just have a wild guess. I am not an Idris implementor (yet). I would say there are two things are happening here. The first the `2` literal gets parsed/repesented as an `Integer` in the expression `x * 2`. Where `x : Double`. Probably an expression is created when copiling such as: `let y = Int 2 in multiply x (castInteger y)`. The second one, `castInteger` implementation in the JS backend creates a syntactically invalid JS snippet.

This
```
return __esPrim_crashExp(invalid cast: + Integer + ' -> ' + Double);
```
should be this:
```
return __esPrim_crashExp('invalid cast:' + 'Integer' + ' -> ' + 'Double');
```

But this is just a wild guess. I didn't spend time invesitgating this issue. If I recall using `* 2.0` works as a workaround.

**Comment by DoctorRyner** [comment]
Well, I hope it'll be fixed

**Comment by gallais** [comment]
Looks like it's caused by [a missing primitive cast](https://github.com/idris-lang/Idris2/blob/3a6e779acff2fe1e73193428122aa197a87bf341/src/Compiler/ES/ES.idr#L311).

If you know how to cast an `IntegerType` (a `BigInt` in javascript AFAIU) to a `Double`
then it should just be a matter of inserting this new line. Probably worth adding all
the other possible casts from Int-like types to Double if you can too.

**Comment by SmiVan** [comment]
I think this issue can now be closed, as I don't get any issues with the code example.
```
yarn run v1.22.10
$ yarn build && yarn test-build && node build/exec/index.js
$ idris2 --build js.ipkg
$ idris2 test/Main.idr --codegen javascript -p js.ipkg -o index.js
{ color: 'red', 'z-index': 9999 }

```

**Comment by gallais** [comment]
Yep, fixed in #961

**Comment by DoctorRyner** [comment]
Thank you for your work!
